### ISSUE TITLE
Add an `injectionNode` option to `loadStyleSheet`

### ISSUE THREAD

**Issue by savetheclocktower** [bazaarvoice/scoutfile#21]
## The problem

The `loadStyleSheet` method injects a `link` tag as a sibling of the first `script` node, which doesn't work on Firebird's container page because we replace everything within the `body` tag soon thereafter.

## Proposed solution

Allow an option called `injectionNode` that lets us specify a different node (in our case the `head`). This lets us customize the behavior of `loadStyleSheet` without changing the behavior for anyone else.

**Comment by brianarn** [comment]
At this point in time, the [current CSS loading strategy](https://github.com/bazaarvoice/scoutfile/blob/985feffb320aee2eb6637806eed7a3697ec2ce41/lib/browser/loader.js#L216) is such:

```javascript
    getFirstScript().parentNode.appendChild(link);
```

There is support for an `options` object, and it's made to be safe at the outset of `loadStyleSheet`, so we should be able to safely look up the `injectionNode` property, falling back to current behavior.

```javascript
    var injectionNode = options.injectionNode || getFirstScript().parentNode;
    injectionNode.appendChild(link);
```

That would use the `injectionNode` if provided, and fall back to the current behavior of looking up the first script element and using its parent.

This approach feels the least disruptive to this project, while supporting our particular use case.

**Comment by brianarn** [comment]
The other approach that had crossed my mind while thinking about this last night, would be to support an `options.injectIntoHead` boolean, and then to do the head lookup here and inject that way, since that's likely going to be the most common use case.

Comparing the two in my mind, the `injectionNode` approach felt less disruptive and offered more flexibility, especially looking forward to scoped styles in the future, where this might be useful to inject into a particular element.

**Comment by rmurphey** [comment]
Whose responsibility is it to make sure that `injectionNode` exists and is a node? Does that fall to the loader code or the consuming application?

**Comment by brianarn** [comment]
IMO, that responsibility should be on the consuming application. As proposed, the consuming application would be making a choice and providing a node, and if they provide something that is invalid, that is on the application. The scout file should be able to safely assume that argument.

It could be guarded with some sort of try/catch or verification that the `appendChild` method exists, but since it's an opt-in behavior, it feels reasonable to me to require proper usage and to not have to worry about guarding against improper use.

**Comment by rmurphey** [comment]
Also, both `loadStyleSheet` and `loadScript` have the same injection behavior; should we offer this option for `loadScript` as well? That seems to make less sense to me, but I wanted to bring it up.

**Comment by lawnsea** [comment]
> Whose responsibility is it to make sure that injectionNode exists and is a node? Does that fall to the loader code or the consuming application?

The consuming application, IMO.

**Comment by lawnsea** [comment]
We could check the `injectionNode` and call back with an error, though. We should probably do that.

**Comment by lawnsea** [comment]
> Also, both loadStyleSheet and loadScript have the same injection behavior; should we offer this option for loadScript as well? That seems to make less sense to me, but I wanted to bring it up.

Yes, I think we should.

**Comment by brianarn** [comment]
It seems like `loadScript` uses a slightly different approach, since it's doing an `insertBefore` as opposed to just appending the child element in. A simple function could be made to append children, which would validate as well, like

```javascript
function appendNode(node, options) {
  var injectionNode = options.injectionNode || getFirstScript().parentNode;
  injectionNode.appendChild(node);
}
```

Then in both `loadScript` and `loadStyleSheet`, it could invoke that method, catching an error if `appendChild` errors out and calling back with that.

**Comment by savetheclocktower** [comment]
My view is that this matters less for `loadScript` because a script still gets evaluated no matter what happens to its `script` node. Hence I can't think of a use case for putting a `script` node in a specific place.

**Comment by rmurphey** [comment]
I'm with @savetheclocktower on `loadScript`, but I don't feel strongly. I guess it won't really hurt anything.

**Comment by savetheclocktower** [comment]
So the points of contention are (1) who handles errors, and (2) whether we do the same thing for `loadScript`. And the consensus seems to be _meh_ on both.

Here are my proposed resolutions:

1. It's easy enough to ensure that `injectionNode` is a node (if present), and throw an error otherwise, so I'm happy to do that.
2. Without an identifiable use case for this option in `loadScript`, I am reluctant to implement it. (If anyone feels strongly otherwise, I'd like to hear their arguments.)

**Comment by brianarn** [comment]
Well, not so much throw an error, but invoke the callback with the error as the first argument, along the lines of how errors are typically handled in callbacks in Node, if I'm understanding the idea correctly.

I have no strong opinion about `loadScript`, just seemed like consistency between the two would be good. That's about as strong an argument as I can make.

**Comment by savetheclocktower** [comment]
Ah. Well, that's another question — do we invoke the callback with an error, or do we throw? If the idea is to make the nature of the error more apparent for developer comfort, seems a throw would be best — it stops the world and the developer knows why as soon as she checks the console.

There's already a situation in `loadStyleSheet` in which we invoke the callback with an error, but that's a timeout error, which is inherently asynchronous. In this case, we'd know right away if `injectionNode` weren't actually a node.

**Comment by rmurphey** [comment]
I'd be inclined to throw. Lots of apps won't use the callback at all.

**Comment by lawnsea** [comment]
Because the stylesheet is injected asynchronously, we will have to call back if the node turns out to be bad. We can (and probably should) also throw if we know the node is bad at call time, of course.

**Comment by savetheclocktower** [comment]
A try/catch around the `appendChild` call would work, right?

**Comment by lawnsea** [comment]
@savetheclocktower yeah, I should already have put that in :)

My point, however, is that we *must* support reporting errors asynchronously, since the `link` is injected on a later event loop turn. We *may* support reporting errors synchronously - and probably should.

**Comment by rmurphey** [comment]
Resolved in #22.
